### PR TITLE
GET users endpoints, seed password hashing

### PR DESCRIPTION
**Pull request by rupol**

### CODE REVIEW

File: auth/auth-router.js [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+const bcrypt = require('bcryptjs');
+const jwt = require('jsonwebtoken');
+const usersModel = require('../users/users-model');
+const secrets = require('../config/secrets');
+
+const router = express.Router();
+
+router.post('/register', async (req, res, next) => {
+  try {
+    const user = await usersModel.add(req.body);
```
Comment (Derrick-Mei) [inline auth/auth-router.js]: Nothing wrong with directly adding the req.body, but I would really like to see you whitelist the needed fields

File: auth/auth-router.js [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+    const passwordValid = await bcrypt.compare(password, user.password);
+
+    if (user && passwordValid) {
+      const token = signToken(user);
+
+      res.status(200).json({
+        message: 'Logged in',
+        user: user.id,
+        token,
+      });
```
Comment (Derrick-Mei) [inline auth/auth-router.js]: I believe this is the pattern that Jason did in class, but I believe there is 1 logical error in it.  If you try to login a user with a username that doesn't exist.  It will successfully return null.  The issue will then be on line 22 with something like `can't get password of undefined` and it will proceed to catch the error instead of saying `invalid credentials...`

To mitigate the problem, you need to have an if/else block and test if user is null first.  Then if user isn't null, you can check password.

From a security standpoint, you would also make the decision when searching the user by username.  If username doesn't exist, do you say `invalid credentials` or do you say `user doesn't exist?`.  I personally would use `user doesn't exist` as it's not as important for this app to be highly secure.  This makes it easier to understand what went wrong for the user.

File: auth/auth-router.js [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+    next(error);
+  }
+});
+
+// this function creates and signs the token
+function signToken(user) {
+  const payload = {
+    subject: user.id,
+    username: user.username,
+    user_type: user.user_type,
```
Comment (Derrick-Mei) [inline auth/auth-router.js]: I have no idea why Jason's example used `subject` as the key name, but I think userId is much more concise here.

File: auth/auth.router.spec.js [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+
+  // does it return the expected data format?
+  expect(res.type).toBe('application/json');
+
+  // does it return the expected data?
+  expect(res.id).toBe(11);
+  expect(res.username).toMatch(/test/i);
+  expect(res.email).toMatch(/test@test.biz/i);
+  expect(res.password).not.toBeNull();
+  expect(res.user_type).toMatch(/admin/i);
```
Comment (Derrick-Mei) [inline auth/auth.router.spec.js]: I love the regex

File: campaigns/campaigns-model.js [hunk 5]
```diff
@@ -0,0 +1,51 @@
+const db = require('../data/config');
+
+function get() {
+  return db('campaigns as c')
+    .join('organizations as o', 'c.org_id', 'o.id')
+    .select('c.*', 'o.name as org_name');
```
Comment (Derrick-Mei) [inline campaigns/campaigns-model.js]: It's not super important here, but shorthand `c.*` to select all from campaign is computationally slower than explicitly writing it out from the server side.  We are talking about millionths of a second for an app this size, so not important for build week.

File: campaigns/campaigns-model.js [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+    .where({ id })
+    .update(changes);
+
+  return getById(id);
+}
+
+function del(id) {
+  return db('campaigns')
+    .where({ id })
+    .del();
```
Comment (Derrick-Mei) [inline campaigns/campaigns-model.js]: I hate delete, because we can't return the deleted item as a confirmation.  I always have to search for the object first to have access to it and then delete it and return it.

File: middleware/authenticate.js [hunk 7]
```diff
@@ -0,0 +1,18 @@
...
+  return (req, res, next) => {
+    try {
+      const token = req.headers.authorization;
+      const decoded = jwt.verify(token, secrets.jwt);
+
+      req.userId = decoded.subject;
+      next();
+    } catch (err) {
+      return res.status(401).json({
+        message: 'User not authenticated. Please log in and try again.',
```
Comment (Derrick-Mei) [inline middleware/authenticate.js]: Nothing logically wrong with your authenticate middleware, but on the post campaign routes this is the only thing stopping any user from posting a new campaign.  What would stop a regular supporter from posting a campaign?  What would stop a organization from posting a campaign to another organization?

File: middleware/authenticate.js [hunk 8]
```diff
@@ -0,0 +1,18 @@
+const jwt = require('jsonwebtoken');
+const secrets = require('../config/secrets');
+
+module.exports = () => {
+  return (req, res, next) => {
+    try {
+      const token = req.headers.authorization;
+      const decoded = jwt.verify(token, secrets.jwt);
+
+      req.userId = decoded.subject;
```
Comment (Derrick-Mei) [inline middleware/authenticate.js]: It's weird, because here Jason changed `subject` back to `userId`

File: organizations/organizations-router.js [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+});
+
+router.get('/:id', async (req, res, next) => {
+  try {
+    const organization = await organizationsModel.getById(req.params.id);
+    res.status(200).json(organization);
+  } catch (err) {
+    next(err);
+  }
+});
```
Comment (Derrick-Mei) [inline organizations/organizations-router.js]: This endpoint also works, but it would be immensely helpful if this endpoint also returned a organization's campaigns.  That way, when a organization user logs in, the front end does not need to hit 1 endpoint to get the org name and then a 2nd endpoint to get the org's campaigns

File: users/users-router.js [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+const authenticate = require('../middleware/authenticate');
+const jwt = require('jsonwebtoken');
+
+const router = express.Router();
+
+router.get('/', authenticate(), async (req, res, next) => {
+  try {
+    const token = req.headers.authorization;
+    const user = jwt.decode(token, { complete: true });
+    const { user_type } = user.payload;
```
Comment (Derrick-Mei) [inline users/users-router.js]: There may be an error here.  `payload` is the name of the object being encrypted, and it is the value that is returned when the token is decoded.  So `user` is technically `payload`.  I think it should be `const { user_type } = user;`

File: users/users-router.js [hunk 11]
```diff
@@ -0,0 +1,44 @@
...
+  } catch (err) {
+    next(err);
+  }
+});
+
+router.get('/:id', authenticate(), async (req, res, next) => {
+  try {
+    const token = req.headers.authorization;
+    const user = jwt.decode(token, { complete: true });
+    const { user_type, subject } = user.payload;
```
Comment (Derrick-Mei) [inline users/users-router.js]: same as above?  You'll need so more checks
